### DISCUSSION TITLE
heroku instructions seem unnecessary

### DISCUSSION THREAD

**Discussion by jrochkind** [ElMassimo/vite_ruby#280] [General]
There are some heroku-specific instructions in the Deploying docs: https://vite-ruby.netlify.app/guide/deployment.html#using-heroku

However, I have found it is not in fact necessary to add the heroku `nodejs` buildpack.  

I can deploy my heroku app that only has the `ruby` buildpack, and everything seems to deploy and vite JS gets built, no problem. In my experience, it does not seem necessary to have any specific instructions for heroku and vite. 

If you *do* add the nodejs buildpack, then it's possible you need to set `NPM_CONFIG_INCLUDE` and/or `YARN_PRODUCTION` as documented. But with just the ruby buildpack, and without setting these values, my app is deploying just fine. 

Eg

```
remote: -----> Preparing app for Rails asset pipeline
remote:        Running: rake assets:precompile
remote:        [Scout] [10/18/22 14:29:00 -0400 run.5775 (1171)] INFO : Scout Agent [5.3.1] Initialized
remote:        [Scout] [10/18/22 14:29:00 -0400 run.5775 (1171)] INFO : Not Loading Instruments
remote:        [Scout] [10/18/22 14:29:00 -0400 run.5775 (1171)] INFO : Scout Agent [5.3.1] Installed
remote:        [Scout] [10/18/22 14:29:00 -0400 run.5775 (1171)] INFO : Monitoring isn't enabled for the [production] environment.
remote:        INFO: ** [Honeybadger] Initializing Honeybadger Error Tracker for Ruby. Ship it! version=4.12.2 framework=rails level=1 pid=1171
remote:        yarn install v1.22.17
remote:        [1/4] Resolving packages...
remote:        [2/4] Fetching packages...
remote:        [3/4] Linking dependencies...
remote:        [4/4] Building fresh packages...
remote:        Done in 8.81s.
remote:        I, [2022-10-18T14:29:20.484775 #1171]  INFO -- : Writing /tmp/build_e831de52/public/assets/application-f9436831173526bb6d7c941a3223cf1a5cee9993b58c10357e2aa9806e787be8.js
remote:        I, [2022-10-18T14:29:20.485076 #1171]  INFO -- : Writing /tmp/build_e831de52/public/assets/application-f9436831173526bb6d7c941a3223cf1a5cee9993b58c10357e2aa9806e787be8.js.gz
remote:        Building with Vite ⚡️
remote:        vite v3.1.8 building for production...
remote:        transforming...
remote:        ✓ 150 modules transformed.

[...]

```

**Comment by ElMassimo** [comment]
Hi Jonathan!

When I wrote these instructions, the `nodejs` buildpack was necessary (don't recall if it was the lack of `node` in the base image or Ruby buildpack, of if it was that the included version was too old to run Vite).

Can you share which base image you are using in Heroku?

That way we can add a note in the docs mentioning that.

**Reply by jrochkind** [reply]
Makes sense. 

I'm on `heroku-20`.  but might make sense to write the docs just like somehow "it probably will just work with the heroku setup you already have, but if it doesn't [ideally here's a sample error message, although that may now be lost to time], you may have to add the nodejs buildpack like so..."

**Reply by jrochkind** [reply]
My guess is that at some point the ruby buildpack has been expanded to work well enough for webpacker (when that was Rails preferred path), which also gave vite what it needed.

**Comment by themire** [comment]
> When precompiling assets in Heroku, it's better to skip pruning of dev dependencies

Why exactly is this? It doesn't seem to make any difference on my staging Heroku app where I've been testing. Just curious to know if I should do it for production!

**Reply by ElMassimo** [reply]
Heroku would install all dependencies in the node buildpack, then prune dev dependencies, to then have Vite Ruby or webpacker install all dependencies again in order to precompile assets. It would be a waste of time.

It won't affect the app though, simply take a little longer (probably imperceptible in a new app with few deps).

**Comment by 9mm** [comment]
It seems preferrable anyway to use the actual buildpack - https://devcenter.heroku.com/articles/ruby-support#installed-binaries

**Reply by jrochkind** [reply]
Yes, heroku started recommending that more explicitly in the past couple months, those docs are fairly newly changed!

**Comment by 9mm** [comment]
Also it seems super confusing as `NPM_CONFIG_INCLUDE` is stated that its needed for "newer" npm versions, yet in the heroku docs it seems I need the other one which you have listed as NPM < 7. so that part was super confusing.

**Comment by 9mm** [comment]
Hey @ElMassimo I just tried and it definitely seems  `NPM_CONFIG_PRODUCTION` is the right one (im using Node 20, NPM 10), the other didn't do anything
```
remote:
remote: -----> Pruning devDependencies
remote:        Skipping because NPM_CONFIG_PRODUCTION is 'false'
```

WIth that said I saw this stuff, I wasn't sure if this is something to worry about?

Sorry to hijack this issue it just seems related

```
remote: -----> Detecting rake tasks
remote: -----> Preparing app for Rails asset pipeline
remote:        Running: rake assets:precompile
remote:        npm WARN config production Use `--omit=dev` instead.
remote:        npm WARN config production Use `--omit=dev` instead.
remote:        npm WARN config production Use `--omit=dev` instead.
remote:        npm WARN config production Use `--omit=dev` instead.
remote:
remote:        added 317 packages, and audited 318 packages in 14s
remote:
remote:        77 packages are looking for funding
remote:          run `npm fund` for details
```
